File: python/src/pmsec/tools/cargo.py
from __future__ import annotations

import re
from pathlib import Path

from pmsec.util.context import Context
from pmsec.util.io import write_atomic
from pmsec.util.lines import read_key, remove_key, set_key
from pmsec.util.paths import cargo_config_path

NAME = "cargo"
KEY = "minimum-release-age"
SECTION = "install"
DOCS = "https://rust-lang.github.io/rfcs/3801-package-cooldown.html"
EXTRAS: list[dict] = []


def path(ctx: Context) -> Path:
    return cargo_config_path(ctx.env, ctx.home)


def _parse_days(value: str | None) -> int | None:
    if value is None:
        return None
    m = re.match(r'^"?\s*(\d+)\s*(d|days?|w|weeks?)\s*"?$', value, re.IGNORECASE)
    if not m:
        return None
    n = int(m.group(1))
    return n * 7 if re.match(r"^(w|weeks?)$", m.group(2), re.IGNORECASE) else n


def read(ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    value = read_key(raw, KEY, section=SECTION)
    return {"path": str(p), "configured": value, "days": _parse_days(value), "extras": []}


def write(days: int, ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    write_atomic(p, set_key(raw, KEY, f'{KEY} = "{days}d"', section=SECTION))
    return {"path": str(p)}


def unset(ctx: Context) -> dict:
    p = path(ctx)
    if not p.exists():
        return {"path": str(p), "removed": False}
    before = p.read_text("utf-8")
    after, removed = remove_key(before, KEY, section=SECTION)
    if removed:
        write_atomic(p, after)
    return {"path": str(p), "removed": removed}

File: node/src/tools/cargo.mjs
import { cargoConfigPath } from "../util/paths.mjs";
import { readSafe, writeAtomic } from "../util/io.mjs";
import { readKey, setKey, removeKey } from "../util/lines.mjs";

export const name = "cargo";
export const key = "minimum-release-age";
export const section = "install";
export const docs = "https://rust-lang.github.io/rfcs/3801-package-cooldown.html";
export const extras = [];

export function path(ctx) { return cargoConfigPath(ctx.env, ctx.home); }

function parseDays(value) {
  if (value === null) return null;
  const m = value.match(/^"?\s*(\d+)\s*(d|days?|w|weeks?)\s*"?$/i);
  if (!m) return null;
  const n = Number(m[1]);
  return /^(w|weeks?)$/i.test(m[2]) ? n * 7 : n;
}

export async function read(ctx) {
  const p = path(ctx);
  const raw = await readSafe(p);
  const value = readKey(raw, key, { section });
  return { path: p, configured: value, days: parseDays(value), extras: [] };
}

export async function write(days, ctx) {
  const p = path(ctx);
  const text = setKey(await readSafe(p), key, `${key} = "${days}d"`, { section });
  await writeAtomic(p, text);
  return { path: p };
}

export async function unset(ctx) {
  const p = path(ctx);
  const before = await readSafe(p);
  const { text: after, removed } = removeKey(before, key, { section });
  if (removed) await writeAtomic(p, after);
  return { path: p, removed };
}
